### ISSUE TITLE
Distributed builds seem to disable local builds

### ISSUE THREAD

**Issue by memberbetty** [NixOS/nix#2589]
I noticed the distributed builds feature, configured it with two remote machines, but while it builds on a single remote machine. It's not using the local machine anymore. 

`nix.buildMachines` is set such that it does not include the local machine and `nix.distributedBuilds =  true; nix.maxJobs = 9;  `. 

Am I missing an option?

 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.87, NixOS, 18.09.git.7e88992 (Jellyfish)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.1.3`
 - channels(root): `"nixos-18.09.1571.7795a7ad5f0"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`

**Comment by tomberek** [comment]
It's possible that remote builds are checked first, and the local machine is only used if connections to the remote machines fail.

**Comment by ryan4yin** [comment]
@tomberek 
Do we have any parameters to define this behavior? For example, local priority, or calculate the priority through the `speedFactor  parameter.

**Comment by jarrodsfarrell** [comment]
Still looking for an answer to this. I've tried this configuration:
```nix
{ config, pkgs, lib, ... }:

let
  buildhost = "buildhost.fqdn.local";
in
{
  nix.buildMachines = [ {
    hostName = "${buildhost}";
    systems = [ "x86_64-linux" "i686-linux" ];
    protocol = "ssh-ng";
    maxJobs = 32;
    speedFactor = 100;
    supportedFeatures = [ "nixos-test" "big-parallel" "kvm" ];
    mandatoryFeatures = [ ];
  } {
    hostName = "localhost";
    systems = [ "x86_64-linux" "i686-linux" ];
    protocol = null;
    maxJobs = 8;
    speedFactor = -100;
    supportedFeatures = [ "nixos-test" "big-parallel" "kvm" ];
    mandatoryFeatures = [ ];
  } ] ;

  nix.distributedBuilds = true;

  nix.extraOptions = ''
    builders-use-substitutes = true
  '';

  nix.settings = {
    trusted-substituters = lib.mkBefore [ "ssh-ng://${buildhost}?priority=1" "https://cache.nixos.org?priority=50"];
    trusted-public-keys = lib.mkBefore [ "buildhost.fqdn.local-1:eCHtUvxostvazNDbUBxnLgWOkcRGQrkfK11PzWRUkGs=" ];
    substituters = lib.mkBefore [ "ssh-ng://${buildhost}" ];
  };

}
```
But it still fails entirely if `buildhost` is not present, instead of using the available `localhost`

**Comment by sellout** [comment]
I’ve run into this as well. I describe my `buildMachines` setup a bit in https://github.com/NixOS/nix/issues/10740#issuecomment-3094723385. It means that my most powerful machine doesn’t do remote builds, and everything else builds “upwards”. When I need to do a local build for some reason, I pass `--builders ""` and the local builder seems to take over fine in that case.

Would love to be able to include the local machine as a peer in `buildMachines`, so that `maxJobs`, `speedFactor`, etc. are taken into account.

**Comment by Gigahawk** [comment]
@sellout in your case the behavior you see is still that all builds are forwarded to a faster machine though right? The machine you trigger a build on never builds anything even if there are unrelated derivations?

**Comment by Gigahawk** [comment]
I've written up a quick repo to demo this issue more repeatably, if anyone manages to get a setup where they can get all available builders (remote and local) to each consume their own maxJobs simultaneously would really appreciate if they showed how they did it.

https://github.com/Gigahawk/parallel_build_test

**Comment by sellout** [comment]
> [@sellout](https://github.com/sellout) in your case the behavior you see is still that all builds are forwarded to a faster machine though right? The machine you trigger a build on never builds anything even if there are unrelated derivations?

That’s correct – haven’t managed both local & distributed builds at the same time.
